c/src/lib.rs

```diff
@@ -533,3 +664,97 @@ impl foxglove::websocket::ServerListener for FoxgloveServerCallbacks {
         }
     }
 }
+
+#[repr(u8)]
+pub enum FoxgloveErrorKind {
+    Unspecified,
+    ValueError,
+    SinkClosed,
+    SchemaRequired,
+    MessageEncodingRequired,
+    ServerAlreadyStarted,
+    Bind,
+    DuplicateChannel,
+    DuplicateService,
+    MissingRequestEncoding,
+    ServicesNotSupported,
+    ConnectionGraphNotSupported,
+    IoError,
+    McapError,
+}
+
+impl From<&foxglove::FoxgloveError> for FoxgloveErrorKind {
+    fn from(error: &foxglove::FoxgloveError) -> Self {
+        match error {
+            foxglove::FoxgloveError::ValueError(_) => FoxgloveErrorKind::ValueError,
+            foxglove::FoxgloveError::SinkClosed => FoxgloveErrorKind::SinkClosed,
+            foxglove::FoxgloveError::SchemaRequired => FoxgloveErrorKind::SchemaRequired,
+            foxglove::FoxgloveError::MessageEncodingRequired => {
+                FoxgloveErrorKind::MessageEncodingRequired
+            }
+            foxglove::FoxgloveError::ServerAlreadyStarted => {
+                FoxgloveErrorKind::ServerAlreadyStarted
+            }
+            foxglove::FoxgloveError::Bind(_) => FoxgloveErrorKind::Bind,
+            foxglove::FoxgloveError::DuplicateChannel(_) => FoxgloveErrorKind::DuplicateChannel,
+            foxglove::FoxgloveError::DuplicateService(_) => FoxgloveErrorKind::DuplicateService,
+            foxglove::FoxgloveError::MissingRequestEncoding(_) => {
+                FoxgloveErrorKind::MissingRequestEncoding
+            }
+            foxglove::FoxgloveError::ServicesNotSupported => {
+                FoxgloveErrorKind::ServicesNotSupported
+            }
+            foxglove::FoxgloveError::ConnectionGraphNotSupported => {
+                FoxgloveErrorKind::ConnectionGraphNotSupported
+            }
+            foxglove::FoxgloveError::IoError(_) => FoxgloveErrorKind::IoError,
+            foxglove::FoxgloveError::McapError(_) => FoxgloveErrorKind::McapError,
+            _ => FoxgloveErrorKind::Unspecified,
+        }
+    }
+}
+
+const FOXGLOVE_ERROR_MAGIC: u32 = 0xDEADDEAD;
+
+#[repr(C)]
+pub struct FoxgloveError {
+    pub kind: FoxgloveErrorKind,
+    magic: u32,
+    pub message: *const c_char,
+}
+
+impl From<foxglove::FoxgloveError> for FoxgloveError {
+    fn from(error: foxglove::FoxgloveError) -> Self {
+        let message = CString::new(error.to_string()).unwrap();
+        Self {
+            // This gives a little protection against people passing an unintialized
+            // FoxgloveError into foxglove_error_free, which is an easy mistake to make in C.
```

Comment:
what makes this mistake easy - is it that we never assign the error if no error was encountered? What if we just assigned a null message (i.e. initialize it) every time there was no error?